default.nix

```diff
@@ -79,4 +84,61 @@ rec {
     in
     lenContent >= lenSuffix
     && builtins.substring (lenContent - lenSuffix) lenContent content == suffix;
+
+  # Determine segment-wise if one path is a prefix of another
+  #
+  # Paths need to be split into segments since a simple string
+  # prefix check would naively conclude things like "/nix/sto"
+  # is a path prefix of "/nix/store" even though they are completely
+  # different directories.
+  #
+  # >>> _isPathPrefix /nix/var/ /nix/var/nix/profiles/
+  # true
+  #
+  # >>> _isPathPrefix "store" /nix/store/
+  # false
+  _isPathPrefix =
+    # Prefix to search for in `path`, can be a string or a path
+    prefixPath:
+    # Path to search, can be a string or a path
+    path:
+    let
+      prefixSegments = _splitPath prefixPath;
+      pathSegments = _splitPath path;
+
+      # Compare the lists of segments. If any segment doesn't match,
+      # the prefix is a not a prefix of the path.
+      #
+      # This iterates over the prefix and should not be run if the
+      # path is shorter than the prefix or it will throw runtime
+      # errors.
+      comparison =
+        builtins.foldl' (acc: prefixSegment:
+          let
+            isPrefix = acc.isPrefix && prefixSegment == builtins.head acc.remainingPath;
+            remainingPath = builtins.tail acc.remainingPath;
+          in
+          { inherit isPrefix remainingPath; }
+        ) { isPrefix = true; remainingPath = pathSegments; } prefixSegments;
+    in
+    # Fail immediately if the path is shorter than the prefix; do not try to
+    # compare.
+    if builtins.length prefixSegments > builtins.length pathSegments
+    then false
+    else comparison.isPrefix;
```

Comment:
Nix is not super fast so I try to think about ways to minimize the number of operations that need to happen. Either by memoizing operations or by using faster builtins.

For example here, for each match, both paths will be split (generating more allocations), then each component of the path will be compared.

If we could assume that the paths are "clean" upfront, which is true in most of the cases, then this could be reduced to a string prefix comparison.

Reply:
Could you elaborate on what you mean by "clean"?

Reply:
Sure, a "clean" path is one that is absolute and doesn't contain unnecessary slashes or dots.

If it's possible to assume that, then it's possible to check that

/home/user/src/project/ is a prefix of /home/user/src/project/src/main.js

Reply:
If I'm not mistaken, Nix paths are clean by default. So, we could either:

a) have the user pass in a path type (enforced via `assert`), or
b) have the matcher automatically turn strings into paths (`root + ("/" + path)`), which would require the matcher to also be passed `root` as a parameter.

I'm more in favor of option (b) here. Much nicer on the user end to write `(inDirectory "src")` than `(inDirectory (root + "/src"))`. Plus, matchers being passed `root` opens them up to being smart about more things.

Reply:
Went with option (b), but can change to (a) or something else if you prefer.

Reply:
Yeah, I was thinking the same yesterday, matchers might want to be initialized with the root.